zellij-server/src/wasm_vm.rs

```diff
@@ -50,77 +54,48 @@ impl From<&PluginInstruction> for PluginContext {
 }
 
 // Thread main --------------------------------------------------------------------------------------------------------
-pub(crate) fn wasm_thread_main(bus: Bus<PluginInstruction>, store: Store, data_dir: PathBuf) {
+pub(crate) fn wasm_thread_main(
+    bus: Bus<PluginInstruction>,
+    store: Store,
+    data_dir: PathBuf,
+    plugins: Plugins,
+) {
     info!("Wasm main thread starts");
+
     let mut plugin_id = 0;
-    let mut plugin_map = HashMap::new();
+    let mut plugin_map = HashMap::<u32, (Instance, PluginEnv)>::new();
+
+    for plugin in plugins.iter() {
+        if let PluginType::Headless = plugin.run {
+            let (instance, plugin_env) =
+                start_plugin(plugin_id, plugin, 0, &bus, &store, &data_dir);
+            plugin_map.insert(plugin_id, (instance, plugin_env));
+            plugin_id += 1;
+        }
+    }
+
     loop {
         let (event, mut err_ctx) = bus.recv().expect("failed to receive event on channel");
         err_ctx.add_call(ContextType::Plugin((&event).into()));
         match event {
-            PluginInstruction::Load(pid_tx, path, tab_index, _allow_exec_host_cmd) => {
-                let plugin_dir = data_dir.join("plugins/");
-                let wasm_bytes = fs::read(&path)
-                    .or_else(|_| fs::read(&path.with_extension("wasm")))
-                    .or_else(|_| fs::read(&plugin_dir.join(&path).with_extension("wasm")))
-                    .unwrap_or_else(|_| panic!("cannot find plugin {}", &path.display()));
-
-                // FIXME: Cache this compiled module on disk. I could use `(de)serialize_to_file()` for that
-                let module = Module::new(&store, &wasm_bytes).unwrap();
-
-                let output = Pipe::new();
-                let input = Pipe::new();
-                let stderr = LoggingPipe::new(
-                    path.as_path().file_name().unwrap().to_str().unwrap(),
-                    plugin_id,
-                );
-                let mut wasi_env = WasiState::new("Zellij")
-                    .env("CLICOLOR_FORCE", "1")
-                    .preopen(|p| {
-                        p.directory(".") // FIXME: Change this to a more meaningful dir
-                            .alias(".")
-                            .read(true)
-                            .write(true)
-                            .create(true)
-                    })
-                    .unwrap()
-                    .stdin(Box::new(input))
-                    .stdout(Box::new(output))
-                    .stderr(Box::new(stderr))
-                    .finalize()
-                    .unwrap();
-
-                let wasi = wasi_env.import_object(&module).unwrap();
-
-                if _allow_exec_host_cmd {
-                    info!("Plugin({:?}) is able to run any host command, this may lead to some security issues!", path);
+            PluginInstruction::Load(pid_tx, run, tab_index) => {
+                if run._allow_exec_host_cmd {
```

Comment:
Yeah the "when" and "where" of starting headless plugins was something I took into consideration. Starting them right after the wasm thread is forked was just the simplest and most obvious choice for a first pass at it. I absolutely could start them via `PluginInstruction::Load` message, but I'm worried it might just obfuscate what's happening. Right after the call to `init_session` in `zellij-server/src/lib.rs` maybe? Is there a reason you want it changed aside from the aesthetics of it being a bit ugly?

In any event I moved the security warning into the call to `start_plugin` so that headless plugins get the warning as well.